### PR TITLE
MapView/Parse backend implementation

### PR DESCRIPTION
**Pull request by s-fukuzawaa**
MapView

- When click on point of interest, marker with window pops up. 
- Clicking the window leads to a compose view page. User can upload image, caption, and date for a post.
- When we back to the map view and click on the marker again, info window with summary of pin pops up. Clicking this leads to detail view of the pin.
- Initial loading loads markers of the Pins that user has stored previously.
- Initial position set to user's current location.

Authentication
- Persistent user login
- User can sign up

Demo
CA2343F2-4BED-4BA3-9807-E645C2F84AF6_2_0_aCA2343F2-4BED-4BA3-9807-E645C2F84AF6_2_0_a

### CODE REVIEW

File: Album/ViewControllers/MapControllers/GoogleMapViewController.m [hunk 1]
```diff
@@ -123,33 +124,110 @@ - (void)mapView:(GMSMapView *)mapView didLongPressAtCoordinate:(CLLocationCoordi
...
+        [markerView.placeNameLabel setText:firstPin[@"placeName"]];
+        // Set date
+        NSString *date = [self.formatter stringFromDate:firstPin[@"traveledOn"]];
+        [markerView.dateLabel setText:date];
+        return markerView;
+    }
+    // Array of pins from the specific coordinate
+    NSArray *pinsFromCoord = [self fetchPinsFromCoord:coordinate];
+    // Check if exisitng pins exist from this coordinate
+    if(pinsFromCoord && pinsFromCoord.count > 0) {
```
Comment (albertomota-codes) [inline Album/ViewControllers/MapControllers/GoogleMapViewController.m:170]: do we really need this? because the `while` will not run if  `i > pinsFromCoord.count`

File: Album/ViewControllers/MapControllers/GoogleMapViewController.m [hunk 2]
```diff
@@ -123,33 +124,110 @@ - (void)mapView:(GMSMapView *)mapView didLongPressAtCoordinate:(CLLocationCoordi
...
+        PFObject *firstPin = pinsFromCoord[0];
+        int i=0;
+        while(i<pinsFromCoord.count) {
```
Comment (albertomota-codes) [inline Album/ViewControllers/MapControllers/GoogleMapViewController.m:173]: instead of the while, we could have do something like this:
https://pinkstone.co.uk/how-to-loop-through-every-element-in-an-nsarray/

File: Album/ViewControllers/MapControllers/GoogleMapViewController.m [hunk 3]
```diff
@@ -123,33 +124,110 @@ - (void)mapView:(GMSMapView *)mapView didLongPressAtCoordinate:(CLLocationCoordi
...
+            PFObject *pin = pinsFromCoord[i];
+            // Save pins into cache data structure
+            if(!self.placeToPins[pin[@"placeName"]]){
+                [self.placeToPins setObject:[[NSMutableArray alloc]init] forKey:pin[@"placeName"]];
```
Comment (albertomota-codes) [inline Album/ViewControllers/MapControllers/GoogleMapViewController.m:177]: why do we need to add an empty array?

File: Album/ViewControllers/MapControllers/GoogleMapViewController.m [hunk 4]
```diff
@@ -123,33 +124,110 @@ - (void)mapView:(GMSMapView *)mapView didLongPressAtCoordinate:(CLLocationCoordi
...
+            [self.placeToPins[pin[@"placeName"]] addObject:pin];
+            // Save images of the specific pin to the cache data structure
+            NSArray* imagesFromPin = [self imagesFromPin:pin.objectId];
+            [self.pinImages setObject:imagesFromPin forKey:pin.objectId];
+            i++;
+        }
+        InfoMarkerView *markerView = [[[NSBundle mainBundle] loadNibNamed:@"InfoExistWindow" owner:self options:nil] objectAtIndex:0];
+        // Set image of the info window to first in the array
+        NSArray* imagesFromPin = self.pinImages[firstPin.objectId];
+        if(imagesFromPin && imagesFromPin.count > 0) {
```
Comment (albertomota-codes) [inline Album/ViewControllers/MapControllers/GoogleMapViewController.m:188]: we checked this `if` in line 170, do we need it again?

File: Album/ViewControllers/MapControllers/GoogleMapViewController.m [hunk 5]
```diff
@@ -161,14 +239,29 @@ - (void)didPost {
...
-    // Get the new view controller using [segue destinationViewController].
-    // Pass the selected object to the new view controller.
     if ([segue.identifier isEqual:@"composeSegue"]){
         ComposeViewController *composeVC = [segue destinationViewController];
         composeVC.placeName = self.infoMarker.title;
         composeVC.coordinate = self.infoMarker.position;
         composeVC.placeID = self.infoMarker.snippet;
         composeVC.delegate = self;
+    }else if([segue.identifier isEqual:@"detailsSegue"]){
+        DetailsViewController *detailsVC = [segue destinationViewController];
```
Comment (albertomota-codes) [inline Album/ViewControllers/MapControllers/GoogleMapViewController.m:249]: I am surprised you did not have to cast.

File: Album/ViewControllers/ProfileControllers/FriendProfileViewController.m [hunk 6]
```diff
@@ -53,27 +53,29 @@ - (void) fetchProfile {
     }
 }
 
+- (void) fetchFriendStatus {
+    
+}
+
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/FriendProfileViewController.m:59]: do we need to mark this as WIP? 
Is this used?

File: Album/ViewControllers/ProfileControllers/SettingsViewController.m [hunk 7]
```diff
@@ -40,13 +40,13 @@ - (IBAction)logoutButton:(id)sender {
...
-}
-*/
+ #pragma mark - Navigation
+ 
+ // In a storyboard-based application, you will often want to do a little preparation before navigation
+ - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+ // Get the new view controller using [segue destinationViewController].
+ // Pass the selected object to the new view controller.
+ }
+ */
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/SettingsViewController.m:50]: i really think we should remove all this, and just add it if required.

File: Album.xcworkspace/xcuserdata/eyeray.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist [hunk 8]
```diff
@@ -36,22 +36,6 @@
             landmarkType = "7">
          </BreakpointContent>
       </BreakpointProxy>
-      <BreakpointProxy
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album.xcworkspace/xcuserdata/eyeray.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist:39]: Addressed in .gitignore but folder still exists. Maybe can be resolved in new branch tomorrow?

File: Album/ViewControllers/AuthenticationControllers/SignUpViewController.m [hunk 9]
```diff
@@ -37,16 +37,14 @@ - (IBAction)didTap:(id)sender {
...
     newUser.username = self.usernameField.text;
     newUser.email = self.emailField.text;
     newUser.password = self.pwField.text;
     newUser[@"profileImage"] = self.profileImageView.file;
-    
-    // call sign up function on the object
+    // Call sign up function on the object
     [newUser signUpInBackgroundWithBlock:^(BOOL succeeded, NSError * error) {
         if (error != nil) {
             NSLog(@"Error: %@", error.localizedDescription);
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/ViewControllers/AuthenticationControllers/SignUpViewController.m:50]: Added alert!

File: Album/ViewControllers/ComposeViewController.m [hunk 10]
```diff
@@ -35,15 +37,14 @@ - (void)viewDidLoad {
     [self.pinImageView setUserInteractionEnabled:YES];
     
     // Set location label
-    self.locationLabel.text = self.placeName;
+    self.locationLabel.text = [self.locationLabel.text stringByAppendingString:self.placeName];
```
[the remark this answers is not in the document]
Comment (s-fukuzawaa) [reply Album/ViewControllers/ComposeViewController.m:40]: The original text has "Location: " as default. Adding the place name will become "Location: San Francisco".

File: Album/ViewControllers/MapControllers/GoogleMapViewController.m (hunk 3 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/MapControllers/GoogleMapViewController.m:177]: This is to add a dictionary entry. I added an empty array to store pins corresponding to a single place.

File: Album/ViewControllers/MapControllers/GoogleMapViewController.m (hunk 5 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/MapControllers/GoogleMapViewController.m:249]: It didn't give me an error haha.

File: Album/ViewControllers/ProfileControllers/FriendProfileViewController.m (hunk 6 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/FriendProfileViewController.m:59]: This is used in the branch I'm working today (7/14)... I guess I added the method before switching to a new branch. Should I remove it?

File: Album/ViewControllers/ProfileControllers/SettingsViewController.m (hunk 7 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/SettingsViewController.m:50]: Got it!

File: Album/ViewControllers/ComposeViewController.m (hunk 10 above)
[the remark this answers is not in the document]
Comment (albertomota-codes) [reply Album/ViewControllers/ComposeViewController.m]: Makes sense. 
Usually, two labels are used. That way, the "Location:" part can have a different font and size.

File: Album/ViewControllers/AuthenticationControllers/SignUpViewController.m [hunk 11]
```diff
@@ -20,129 +20,129 @@ @interface SignUpViewController () <UINavigationControllerDelegate, UIImagePicke
...
-    }
-    NSData *imageData = UIImagePNGRepresentation(image);
-    // Get image data and check if that is not nil
-    if (!imageData) {
-        return nil;
-    }
-    return [PFFileObject fileObjectWithName:@"image.png" data:imageData];
+	// Check if image is not nil
+	if (!image) {
+		return nil;
```
Comment (albertomota-codes) [inline Album/ViewControllers/AuthenticationControllers/SignUpViewController.m:127]: why the change in indentation? 
Did you run XCFormat?

**Review by albertomota-codes — APPROVED** [review]
